cou/steps/hypervisor.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2024 Canonical Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Hypervisor planner."""
+import logging
+from dataclasses import dataclass
+
+from cou.apps.base import ApplicationUnit, OpenStackApplication
+from cou.steps import (
+    HypervisorUpgradePlan,
+    PostUpgradeStep,
+    PreUpgradeStep,
+    UpgradePlan,
+)
+from cou.utils.openstack import OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Hypervisor:
+    """Hypervisor containing units for multiple applications."""
+
+    name: str
+    units: list[ApplicationUnit]
+
+
+@dataclass(frozen=True)
+class AZ:
```

Comment:
I don't think naming AZ here is good because an AZ is more than just a collection of hypervisors. Maybe we can call something else like `HypervisorsGrouping` or `HypervisorsCollection`.

If for some reason in the future someone wants to group the Hypervisors by a different algorithm like just for empty VMs, this class will still  have a representative name.

Reply:
That sound like good idea, since both AZ and the Hypervisors are just placeholder for grouped objects, we could easily used just `list[lists[lists[ApplicationUnit]]]`. In next PR, where I will introduce the functionality of grouping we can change this names (potentially use  also Machine class instead of Hypervisor).

These objects are just for better imagination for now, we can discuss further comment in PR we they will be used.